mitiq/folding_cirq.py

```diff
@@ -1,76 +1,234 @@
+
+    Args:
+        circuit: Circuit to fold.
+        moment_indices: Indices of moments with gates to be folded.
+        gate_indices: Specifies which gates within each moment to fold.
+
+    Examples:
+        (1) Folds the first three gates in moment two.
+        >>> fold_gates(circuit, moment_indices=[1], gate_indices=[(0, 1, 2)])
+
+        (2) Folds gates with indices 1, 4, and 5 in moment 0,
+            and gates with indices 0, 1, and 2 in moment 1.
+        >>> fold_gates(circuit, moment_indices=[0, 3], gate_indices=[(1, 4, 5), (0, 1, 2)])
+    """
+    folded = deepcopy(circuit)
+    moment_index_shift = 0
+    for (i, moment_index) in enumerate(moment_indices):
+        folded = fold_gates_in_moment(folded, moment_index + moment_index_shift, gate_indices[i])
+        moment_index_shift += 2 * len(gate_indices[i])  # Folding gates adds moments
+    return folded
+
+
+def fold_moments(circuit: Circuit, moment_indices: List[int]) -> Circuit:
+    """Returns a new circuit with moments folded by mapping
+
+    M_i -> M_i M_i^dag M_i
+
+    where M_i is a moment specified by an integer in moment_indices.
+
+    Args:
+        circuit: Circuit to apply folding operation to.
+        moment_indices: List of integers that specify moments to fold.
+    """
+    folded = deepcopy(circuit)
+    shift = 0
+    for i in moment_indices:
+        folded.insert(i + shift, [circuit[i], inverse(circuit[i])])
+        shift += 2
+    return folded
+
+
+def fold_gates_from_left(circuit: Circuit, stretch: float) -> Circuit:
+    """Returns a new folded circuit by applying the map G -> G G^dag G to a subset of gates  of the input circuit.
+
+    The folded circuit has a number of gates approximately equal to stretch * n where n is the number of gates in
+    the input circuit.
+
+    Args:
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by. Any real number in the interval [1, 3].
+
+    Note:
+        Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
+    """
+    folded = deepcopy(circuit)
+
+    if np.isclose(stretch, 1.0, atol=1e-2):
+        return folded
+
+    if not 1 < stretch <= 3:
+        raise ValueError("The stretch factor must be a real number between 1 and 3.")
 
-    if not ((stretch >= 1) and (stretch <= 3)):
-        raise ValueError("The stretch factor must be a real number within 1 and 3.")
+    ngates = len(list(folded.all_operations()))
+    num_to_fold = int(ngates * (stretch - 1.0) / 2.0)
+    num_folded = 0
+    moment_shift = 0
 
-    depth = len(circuit)
-    reduced_depth = int(depth * (stretch - 1) / 2)
-    sub_indices = random.sample(range(depth), reduced_depth)
-    return fold_gates(circuit, sub_indices)
+    for (moment_index, moment) in enumerate(circuit):
+        for gate_index in range(len(moment)):
+            # TODO: It could be expensive to call fold_gate...(...) which makes a deepcopy of the circuit each call.
```

Comment:
Any reason not to update this to use the in place `_fold_gate_at_index_in_moment` and remove the todo?

Reply:
Nope -- missed this, thanks!